### ISSUE TITLE
V4: Compile to shared library

### ISSUE THREAD

**Issue by jckgore** [aws/aws-iot-device-sdk-embedded-C#1289]
I have a need to compile the latest version as a shared library for a Linux-based industrial controller in order to reference it from a higher level, proprietary language (LabVIEW). Ideally, I'd also like to compile to a Windows DLL for using in debugging and other projects as well.

It looks like this used to be supported (#51, #414, #439, #560) but there's no instructions for use and I'm pretty terrible at CMake.

So the ask is:
How do I compile into a Linux shared library (\*.so)?
How do I compile into a Windows shared library (\*.dll)?

**Comment by sarenameas** [comment]
Hello again @jckgore,

We no longer have support through CMake for creating shared libraries from the aws-iot-device-sdk-embedded-C. I have a question about the past support you link to. Did that used to create a .dll for each library or one monolithic one? 

For each of libraries

**Comment by jckgore** [comment]
To be honest, I have no idea. I haven't used that old version, I just saw it when searching through Github before opening an issue.

It looks like #560 was a PR into the v4beta - I was hoping that meant it would make it into the latest version.

Is a shared library something you would consider supporting? With how common it is to use RPi and other cheap Linux devices for IoT it would be nice if static compilation wasn't required.

**Comment by sarenameas** [comment]
My first post was accidentally posted too early and I didn't get to finish lol. But I will just write a new post.

My colleague showed me how the binaries looks on an older build from last year. Each of the libraries are their own .so. So you have multiple, one for each library.

We don't currently have support to build all libraries at once statically in our CMake system because each customers has different toolchains and most importantly needs for compile time configurations (how long some timeouts are and how large some arrays are to tune for their system). Because of this we have decided to simply present the files needed for each of the libraries all under a single folder. Also since our libraries individually do not have any dependencies outside of their submodule, it is easier than ever to get all the files needed. 

Each library is a submodule under: **libraries/standard/** and **libraries/aws/**. All of the source code needed to compile a single library is under the **source/** folder for that library. The README.md for each library outlines the need for a config file or not and any other global preprocessor macros needed. 

So long story short you will need to bring the **/source** files into your IDE, makefile, or other build tool and generate the shared library. 
Does this help?

**Comment by sarenameas** [comment]
Hello again @jckgore ,

Do you have any questions? Feel free to open a new issue or reopen this issue if you have any other concerns.

Thanks!
Sarena
